File: internal/view/report.go
// Package view turns the collector's state into the two views: versioned JSON
// for an agent and console text for a person.
package view

import (
	"time"

	"github.com/neoromantic/ai-usage/internal/collect"
	"github.com/neoromantic/ai-usage/internal/snapshot"
	"github.com/neoromantic/ai-usage/internal/state"
	"github.com/neoromantic/ai-usage/internal/team"
)

// SchemaVersion is the agent JSON contract. A field may change meaning only
// with a new version.
const SchemaVersion = 3

const (
	// StaleAfter marks a quota reading as old in both views.
	StaleAfter = 6 * time.Hour
	// SilentAfter marks a device that has not reported for a day.
	SilentAfter = 24 * time.Hour
)

// Report is everything both views show. The console draws it as one page:
// what needs attention, the subscriptions, the devices against the
// subscriptions, and this device's projects.
type Report struct {
	SchemaVersion int       `json:"schema_version"`
	GeneratedAt   time.Time `json:"generated_at"`
	Collector     Collector `json:"collector"`
	// Attention is what needs attention now, most urgent first.
	Attention []Attention `json:"attention"`
	// Providers are this device's sources and accounts.
	Providers []Provider `json:"providers"`
	// Projects are this device's projects over every account, most tokens
	// in the last 7 days first.
	Projects []Project `json:"projects"`
	Team     Team      `json:"team"`
}

type Collector struct {
	Version       string     `json:"version"`
	Device        string     `json:"device"`
	DeviceLabel   string     `json:"device_label"`
	OSUser        string     `json:"os_user"`
	Team          string     `json:"team"`
	LastRunAt     *time.Time `json:"last_run_at"`
	LastSuccessAt *time.Time `json:"last_success_at"`
	LastError     *string    `json:"last_error"`
	LastErrorAt   *time.Time `json:"last_error_at"`
	Relay         Relay      `json:"relay"`
	Schedule      Schedule   `json:"schedule"`
	Update        Update     `json:"update"`
}

type Relay struct {
	URL        *string    `json:"url"`
	LastPushAt *time.Time `json:"last_push_at"`
	LastPullAt *time.Time `json:"last_pull_at"`
	Pending    bool       `json:"pending"`
	LastError  *string    `json:"last_error"`
}

type Schedule struct {
	Registered bool    `json:"registered"`
	Foreground bool    `json:"foreground"`
	Error      *string `json:"error"`
}

type Update struct {
	CheckedAt *time.Time `json:"checked_at"`
	Latest    *string    `json:"latest"`
	Staged    *string    `json:"staged"`
	Error     *string    `json:"error"`
}

// Attention kinds, most urgent first.
const (
	AttentionOut    = "out"    // a window is at 100%
	AttentionOver   = "over"   // a window will run out before it resets
	AttentionError  = "error"  // a device's collector or one of its harnesses fails
	AttentionSilent = "silent" // a device has not reported for a day
	AttentionOld    = "old"    // devices run an older release
	AttentionUnder  = "under"  // past half of a window, its forecast is under 50%
)

// Attention is one thing that needs attention. The fields that apply depend
// on Kind.
type Attention struct {
	Kind string `json:"kind"`
	// Provider, Account, and Name are the account of an out, over, or under
	// window: its label and its short name.
	Provider string `json:"provider,omitempty"`
	Account  string `json:"account,omitempty"`
	Name     string `json:"name,omitempty"`
	// Window names the window when it is not the account's main one.
	Window string `json:"window,omitempty"`
	// Devices is the device of an error or silence, or every device on an
	// old release.
	Devices []string `json:"devices,omitempty"`
	// At is when an out window resets, when an over window runs out, or
	// when a silent device last reported.
	At *time.Time `json:"at,omitempty"`
	// ResetsAt is when an over or under window resets.
	ResetsAt *time.Time `json:"resets_at,omitempty"`
	// Percent is an over or under window's forecast at its reset.
	Percent *float64 `json:"percent,omitempty"`
	// ReadingAge is how old the reading is, in seconds, when it is stale.
	ReadingAge int64 `json:"reading_age_seconds,omitempty"`
	// Message is an error's text, or the newest release for old devices.
	Message string `json:"message,omitempty"`
}

type Provider struct {
	Provider string    `json:"provider"`
	Status   string    `json:"status"`
	Error    *string   `json:"error"`
	Homes    []string  `json:"homes"`
	Accounts []Account `json:"accounts"`
}

// Account is one account on this device.
type Account struct {
	Label string `json:"label"`
	// Name is the account's short name: the one the team gave it with
	// `ai-usage alias`, else the part of an email before the @, or the first
	// 8 characters of an id. Two that would be the same within a provider are
	// both the full label.
	Name    string `json:"name"`
	Current bool   `json:"current"`
	// Home is the harness home the account is logged in to now.
	Home string  `json:"home,omitempty"`
	Plan *string `json:"plan"`
	// State is the worst state of the account's windows that limit it: out,
	// over, tight, ok, or under, else unknown.
	State        string          `json:"state"`
	Quota        *Quota          `json:"quota"`
	Link         *Link           `json:"link"`
	Sessions     int             `json:"sessions"`
	Tokens       snapshot.Tokens `json:"tokens"`
	Usage        Usage           `json:"usage"`
	Days         []int64         `json:"days"`
	LinkedUsage  []LinkedUsage   `json:"linked_usage"`
	LastActiveAt *time.Time      `json:"last_active_at"`
	Projects     []Project       `json:"projects"`
}

// Link names the account of another provider an account is assumed to bill
// through. On a team account the label is empty when the reading it shares
// matched no account, or several, on the device that sent it.
type Link struct {
	Provider string `json:"provider"`
	Label    string `json:"label"`
}

// LinkedUsage is what an account of another provider spent through this one.
// Those tokens are that account's, never added to this one's. A team entry
// also names that account and the devices it ran on.
type LinkedUsage struct {
	Provider string          `json:"provider"`
	Label    string          `json:"label,omitempty"`
	Devices  []string        `json:"devices,omitempty"`
	Sessions int             `json:"sessions"`
	Tokens   snapshot.Tokens `json:"tokens"`
}

// Quota is an account's reading. In the team view each window is the newest
// reading of it from any device, so windows can differ in age.
type Quota struct {
	// ObservedAt is the newest window's reading time.
	ObservedAt time.Time `json:"observed_at"`
	AgeSeconds int64     `json:"age_seconds"`
	Stale      bool      `json:"stale"`
	Source     string    `json:"source,omitempty"`
	// From names the provider whose linked account took the reading.
	From string `json:"from,omitempty"`
	// Device is the device whose reading is newest.
	Device  string   `json:"device,omitempty"`
	Windows []Window `json:"windows"`
}

// Forecast states. A window at 100% is out; otherwise the state is the band
// its forecast falls in. Unknown is a window with no forecast: it reset
// since it was read, its length or reset time is unknown, or less than a
// tenth of it has passed and it is not over.
const (
	StateOut     = "out"
	StateOver    = "over"
	StateTight   = "tight"
	StateOK      = "ok"
	StateUnder   = "under"
	StateUnknown = "unknown"
)

type Window struct {
	Name     string     `json:"name"`
	Percent  float64    `json:"percent"`
	ResetsAt *time.Time `json:"resets_at"`
	Minutes  int        `json:"minutes,omitempty"`
	// Main is the account's main window, the weekly one where it has one.
	Main       bool      `json:"main"`
	ObservedAt time.Time `json:"observed_at"`
	// Stale is a reading older than 6 hours.
	Stale bool `json:"stale"`
	// Reset says the window reset since it was read, so how full it is now
	// is not known.
	Reset    bool      `json:"reset"`
	State    string    `json:"state"`
	Forecast *Forecast `json:"forecast"`
}

// Forecast is how full a window will be at its reset if it is used from now
// on at its average pace so far.
type Forecast struct {
	// Percent is the forecast, as a whole percent up to 999.
	Percent float64 `json:"percent"`
	// Elapsed is the share of the window that had passed at the reading,
	// from 0 to 1.
	Elapsed float64 `json:"elapsed"`
	// RunsOutAt is when the window reaches 100% at that pace, when the
	// forecast is over 100%.
	RunsOutAt *time.Time `json:"runs_out_at"`
}

// Usage is input plus output tokens, cache left out, in the report's
// periods. Days are UTC days and each period ends with the report's: today
// is the report's UTC day, and 7d, 30d, and 90d are that many UTC days up to
// and including it.
type Usage struct {
	Today   int64 `json:"today"`
	Week    int64 `json:"7d"`
	Month   int64 `json:"30d"`
	Quarter int64 `json:"90d"`
}

// Project is usage in one working directory. On this device's list it is
// every account's; under an account, that account's.
type Project struct {
	Path     string          `json:"path"`
	Sessions int             `json:"sessions"`
	Tokens   snapshot.Tokens `json:"tokens"`
	Usage    Usage           `json:"usage"`
	// Providers are the harnesses that used it, most tokens first.
	Providers    []string   `json:"providers,omitempty"`
	LastActiveAt *time.Time `json:"last_active_at"`
}

type Team struct {
	PulledAt *time.Time `json:"pulled_at"`
	// Latest is the newest collector release any device in the team runs,
	// or the newest this device's update check saw, whichever is newer.
	Latest    *string        `json:"latest_version"`
	Devices   []TeamDevice   `json:"devices"`
	Providers []TeamProvider `json:"providers"`
	Matrix    Matrix         `json:"matrix"`
}

type TeamDevice struct {
	Device           string     `json:"device"`
	Label            string     `json:"label"`
	OSUser           string     `json:"os_user"`
	This             bool       `json:"this_device"`
	CollectorVersion string     `json:"collector_version"`
	CollectedAt      time.Time  `json:"collected_at"`
	AgeSeconds       int64      `json:"age_seconds"`
	LastSuccessAt    *time.Time `json:"last_success_at"`
	LastError        *string    `json:"last_error"`
	Sources          []Source   `json:"sources"`
	// Error is what fails on the device now: a source's error, named after
	// its provider, else the last run's error when it is newer than the
	// last success.
	Error *string `json:"error"`
	// Silent is a device that has not reported for a day.
	Silent bool `json:"silent"`
	// Old is a device on an older release than the team's newest.
	Old   bool  `json:"old"`
	Usage Usage `json:"usage"`
}

type Source struct {
	Provider string  `json:"provider"`
	Status   string  `json:"status"`
	Error    *string `json:"error"`
}

type TeamProvider struct {
	Provider string `json:"provider"`
	// Accounts are in the order the report lists them: the worst state
	// first (out, over, tight, ok, under, then unknown), ties to the one
	// with less left, then by label.
	Accounts []TeamAccount `json:"accounts"`
}

// TeamAccount adds tokens across devices. Each window of the quota is the
// newest reading any device has of it; percentages are never added.
type TeamAccount struct {
	Label string `json:"label"`
	Name  string `json:"name"`
	// Alias is the name the team gave the account, when it has one.
	Alias *string `json:"alias"`
	// Subscription is an account with a quota of its own: Claude, Codex, and
	// Grok. Hermes is a harness: what it spends through a login is that
	// login's, and its other accounts, such as API keys, have no quota.
	Subscription bool `json:"subscription"`
	// Current says the account is logged in on this device.
	Current bool     `json:"current"`
	Devices []string `json:"devices"`
	Plan    *string  `json:"plan"`
	State   string   `json:"state"`
	Quota   *Quota   `json:"quota"`
	Link    *Link    `json:"link"`
	// Sessions and Tokens are over the 90 days the devices keep.
	Sessions int             `json:"sessions"`
	Tokens   snapshot.Tokens `json:"tokens"`
	Usage    Usage           `json:"usage"`
	// Users counts the devices with tokens on the account since its main
	// window began, or in the last 7 days when it has none, what linked
	// accounts spent through it included. Busiest is the one with the most.
	Users        int           `json:"users"`
	Busiest      *string       `json:"busiest"`
	LastActiveAt *time.Time    `json:"last_active_at"`
	PerDevice    []DeviceUsage `json:"per_device"`
	LinkedUsage  []LinkedUsage `json:"linked_usage"`
}

// DeviceUsage is one device's share of a team account.
type DeviceUsage struct {
	Device       string          `json:"device"`
	DeviceID     string          `json:"device_id"`
	Current      bool            `json:"current"`
	Sessions     int             `json:"sessions"`
	Tokens       snapshot.Tokens `json:"tokens"`
	Usage        Usage           `json:"usage"`
	LastActiveAt *time.Time      `json:"last_active_at"`
}

// Matrix is who spends what: every team device against every subscription,
// and the tokens that have no subscription.
type Matrix struct {
	// Columns are the subscriptions, grouped by provider in the order of
	// the team's providers and accounts, then one column per provider for
	// the tokens with no quota.
	Columns []Column `json:"columns"`
	// Rows are the devices, the most tokens in the last 7 days first.
	Rows []Row `json:"rows"`
}

type Column struct {
	Provider string `json:"provider"`
	// Label is the subscription's account, empty in a no-quota column.
	Label   string `json:"label,omitempty"`
	Name    string `json:"name"`
	NoQuota bool   `json:"no_quota"`
	State   string `json:"state"`
	// Percent is how full the subscription's main window is, when that is
	// known; the share mode splits it between the devices.
	Percent *float64 `json:"percent"`
	Usage   Usage    `json:"usage"`
	// WindowTokens is the team's tokens since the main window began.
	WindowTokens int64 `json:"window_tokens"`
}

type Row struct {
	Device   string `json:"device"`
	DeviceID string `json:"device_id"`
	// Cells has one entry per column.
	Cells []Cell `json:"cells"`
	Usage Usage  `json:"usage"`
}

type Cell struct {
	Usage Usage `json:"usage"`
	// WindowTokens is the device's tokens since the column's main window
	// began.
	WindowTokens int64 `json:"window_tokens"`
	// Share estimates how much of the column's window the device used, in
	// percent: its tokens since the window began over the team's, times how
	// full the window is. A column's shares add up to its Percent.
	Share *float64 `json:"share"`
}

// Input is everything a report is built from.
type Input struct {
	Version  string
	RelayURL string
	Config   state.Config
	State    *state.State
	Key      *team.Key
	Doc      snapshot.Doc
	Team     collect.TeamCache
	Hostname string
	OSUser   string
	Now      time.Time
}

File: macos/Sources/AIUsageKit/Report.swift
import Foundation

/// The report `ai-usage report --json` prints, schema version 4, as
/// docs/json-schema.md describes it. Only the fields the app shows are here;
/// the decoder ignores the rest.
public struct Report: Decodable, Sendable {
    public static let schemaVersion = 4

    public let schemaVersion: Int
    public let generatedAt: Date
    public let collector: Collector
    public let attention: [Attention]
    public let providers: [Provider]
    public let projects: [Project]
    public let team: Team

    /// Decodes a report, or says which schema version it has when that is
    /// not the one this app reads.
    public static func decode(_ data: Data) throws -> Report {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .custom { d in
            let c = try d.singleValueContainer()
            let s = try c.decode(String.self)
            guard let date = parseTime(s) else {
                throw DecodingError.dataCorruptedError(in: c, debugDescription: "not an RFC 3339 time: \(s)")
            }
            return date
        }
        struct Version: Decodable { let schemaVersion: Int }
        let version = try decoder.decode(Version.self, from: data).schemaVersion
        guard version == schemaVersion else { throw ReportError.schema(version) }
        return try decoder.decode(Report.self, from: data)
    }
}

public enum ReportError: Error, LocalizedError, Equatable {
    case schema(Int)

    public var errorDescription: String? {
        switch self {
        case .schema(let v):
            return "This app reads report schema \(Report.schemaVersion); ai-usage printed \(v). Update the app and ai-usage to the same release."
        }
    }
}

/// Parses an RFC 3339 time with or without fractional seconds. The
/// fraction can have up to nine digits, which ISO8601DateFormatter does not
/// read, so it is split off and added back.
public func parseTime(_ s: String) -> Date? {
    var base = Substring(s)
    var fraction = 0.0
    if let dot = s.firstIndex(of: ".") {
        let digits = s[s.index(after: dot)...].prefix(while: \.isNumber)
        guard !digits.isEmpty else { return nil }
        fraction = Double("0." + digits) ?? 0
        base = s[..<dot] + s[digits.endIndex...]
    }
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime]
    return f.date(from: String(base))?.addingTimeInterval(fraction)
}

public struct Collector: Decodable, Sendable {
    public let version: String
    public let device: String
    public let deviceLabel: String
    public let osUser: String
    public let team: String
    public let lastRunAt: Date?
    public let lastSuccessAt: Date?
    public let lastError: String?
    public let lastErrorAt: Date?
    public let relay: Relay
    public let schedule: Schedule
    public let update: Update
    /// The collection, the relay, and the update, as the console's header
    /// shows them; empty from an ai-usage that does not say.
    @Absent public var health: [Health]

    public struct Relay: Decodable, Sendable {
        public let url: String?
        public let lastPushAt: Date?
        public let lastPullAt: Date?
        public let pending: Bool
        public let lastError: String?
    }

    public struct Schedule: Decodable, Sendable {
        public let registered: Bool
        public let foreground: Bool
        public let error: String?
    }

    public struct Update: Decodable, Sendable {
        public let checkedAt: Date?
        public let latest: String?
        public let staged: String?
        public let error: String?
    }

    /// How one part of the collector is doing.
    public struct Health: Decodable, Sendable {
        /// `collection`, `relay`, or `update`.
        public let item: String
        /// How the item is, such as `failed` or `pending`; `ok` when nothing
        /// is wrong with it.
        public let status: String
        /// `ok`, `warn`, `error`, `off`, or `info`: the color of its dot.
        public let state: String
        /// When it last collected, or failed to; when it last reached the
        /// relay; when it last checked for a release.
        public let at: Date?
        /// The staged or the available release.
        public let release: String?
    }
}

/// A field that ai-usage added within schema version 4: an earlier release
/// leaves it out, and it reads as empty, false, or 0.
@propertyWrapper
public struct Absent<Value: Decodable & Sendable & AbsentValue>: Decodable, Sendable {
    public var wrappedValue: Value

    public init(wrappedValue: Value) {
        self.wrappedValue = wrappedValue
    }

    public init(from decoder: Decoder) throws {
        wrappedValue = try decoder.singleValueContainer().decode(Value.self)
    }
}

/// What a field that is not there reads as.
public protocol AbsentValue {
    static var absent: Self { get }
}

extension Bool: AbsentValue {
    public static var absent: Bool { false }
}

extension Int: AbsentValue {
    public static var absent: Int { 0 }
}

extension Array: AbsentValue {
    public static var absent: [Element] { [] }
}

extension KeyedDecodingContainer {
    public func decode<Value>(_: Absent<Value>.Type, forKey key: Key) throws -> Absent<Value> {
        Absent(wrappedValue: try decodeIfPresent(Value.self, forKey: key) ?? .absent)
    }
}

public struct Attention: Decodable, Sendable {
    public enum Kind: String, Decodable, Sendable {
        case out, over, error, silent, old, under, other

        public init(from decoder: Decoder) throws {
            self = Kind(rawValue: try decoder.singleValueContainer().decode(String.self)) ?? .other
        }
    }

    public let kind: Kind
    public let provider: String?
    public let account: String?
    public let name: String?
    public let window: String?
    public let devices: [String]?
    public let at: Date?
    public let resetsAt: Date?
    public let percent: Double?
    public let readingAgeSeconds: Int?
    public let message: String?
}

public struct Provider: Decodable, Sendable {
    public let provider: String
    public let status: String
    public let error: String?
    public let homes: [String]
    public let accounts: [Account]
}

/// One account on this device.
public struct Account: Decodable, Sendable {
    public let label: String
    public let name: String
    public let current: Bool
    public let plan: String?
    public let state: String
    public let quota: Quota?
    public let link: Link?
    public let sessions: Int
    public let usage: Usage
    public let lastActiveAt: Date?
}

public struct Link: Decodable, Sendable {
    public let provider: String
    public let label: String
}

public struct Quota: Decodable, Sendable {
    public let observedAt: Date
    public let ageSeconds: Int
    public let stale: Bool
    public let source: String?
    public let from: String?
    public let device: String?
    public let windows: [Window]

    public var main: Window? { windows.first(where: \.main) }

    /// A window's name beside the main one: "7d Fable" is "Fable" when the
    /// main window is "7d".
    public func shortName(_ window: String) -> String {
        guard let m = main?.name, window.hasPrefix(m + " ") else { return window }
        return String(window.dropFirst(m.count + 1))
    }

    public struct Window: Decodable, Sendable {
        public let name: String
        public let percent: Double
        public let resetsAt: Date?
        public let minutes: Int?
        public let main: Bool
        /// The window limits the account: it is the main window, or one
        /// that limits the account more. False from an ai-usage that does
        /// not say.
        @Absent public var limits: Bool
        public let observedAt: Date?
        public let stale: Bool
        public let reset: Bool
        public let unread: Bool?
        public let state: String
        public let forecast: Forecast?

        /// How full the window is now is known: it has not reset since it was
        /// read, and the reading covers it.
        public var known: Bool { !reset && unread != true }
    }
}

public struct Forecast: Decodable, Sendable {
    public let percent: Double
    public let elapsed: Double
    public let runsOutAt: Date?
}

/// Input plus output tokens in the report's four periods.
public struct Usage: Decodable, Sendable {
    public let today: Int
    public let week: Int
    public let month: Int
    public let quarter: Int

    enum CodingKeys: String, CodingKey {
        case today, week = "7d", month = "30d", quarter = "90d"
    }

    public subscript(p: Period) -> Int {
        switch p {
        case .today: return today
        case .week: return week
        case .month: return month
        case .quarter: return quarter
        }
    }
}

/// A part of a whole in percent, in each period; nil where the whole is none.
public struct Share: Decodable, Sendable {
    public let today: Double?
    public let week: Double?
    public let month: Double?
    public let quarter: Double?

    enum CodingKeys: String, CodingKey {
        case today, week = "7d", month = "30d", quarter = "90d"
    }

    public subscript(p: Period) -> Double? {
        switch p {
        case .today: return today
        case .week: return week
        case .month: return month
        case .quarter: return quarter
        }
    }
}

public enum Period: String, CaseIterable, Identifiable, Sendable {
    case today, week, month, quarter

    public var id: Self { self }

    public var title: String {
        switch self {
        case .today: return "Today"
        case .week: return "7 Days"
        case .month: return "30 Days"
        case .quarter: return "90 Days"
        }
    }
}

public struct Tokens: Decodable, Sendable {
    public let input: Int
    public let output: Int
    public let cacheRead: Int
    public let cacheWrite: Int
}

/// A git repository, or a folder outside any, with the sessions in its
/// subfolders and worktrees.
public struct Project: Decodable, Sendable {
    public let path: String
    /// How many working folders count under the project, 1 or more; 0 from
    /// an ai-usage that does not say.
    @Absent public var folders: Int
    public let sessions: Int
    public let usage: Usage
    public let providers: [String]?
    public let lastActiveAt: Date?
}

public struct Team: Decodable, Sendable {
    public let pulledAt: Date?
    public let latestVersion: String?
    public let devices: [TeamDevice]
    public let providers: [TeamProvider]
    public let matrix: Matrix
}

public struct TeamDevice: Decodable, Sendable {
    public let device: String
    public let label: String
    public let osUser: String
    public let thisDevice: Bool
    public let collectorVersion: String
    public let collectedAt: Date?
    public let ageSeconds: Int
    public let lastSuccessAt: Date?
    public let lastError: String?
    public let updateError: String?
    public let sources: [Source]
    public let error: String?
    public let silent: Bool
    public let old: Bool
    public let behindSince: Date?
    /// An old device that does not update itself. False from an ai-usage
    /// that does not say.
    @Absent public var notUpdating: Bool
    public let usage: Usage

    public struct Source: Decodable, Sendable {
        public let provider: String
        public let status: String
        public let error: String?
    }
}

public struct TeamProvider: Decodable, Sendable {
    public let provider: String
    public let accounts: [TeamAccount]
}

/// An account summed across the team's devices.
public struct TeamAccount: Decodable, Sendable {
    public let label: String
    public let name: String
    public let alias: String?
    public let subscription: Bool
    public let current: Bool
    public let devices: [String]
    public let plan: String?
    public let state: String
    public let quota: Quota?
    public let link: Link?
    public let sessions: Int
    public let usage: Usage
    public let users: Int
    public let busiest: String?
    public let lastActiveAt: Date?
    public let perDevice: [PerDevice]
    public let linkedUsage: [LinkedUsage]

    public struct PerDevice: Decodable, Sendable {
        public let device: String
        public let current: Bool
        public let sessions: Int
        public let usage: Usage
        public let lastActiveAt: Date?
    }

    public struct LinkedUsage: Decodable, Sendable {
        public let provider: String
        public let label: String?
        public let devices: [String]?
        public let sessions: Int
        public let tokens: Tokens
    }
}

public struct Matrix: Decodable, Sendable {
    public let columns: [Column]
    public let rows: [Row]

    public struct Column: Decodable, Sendable {
        public let provider: String
        public let label: String?
        public let name: String
        public let noQuota: Bool
        public let state: String
        public let percent: Double?
        public let usage: Usage
        public let windowTokens: Int
    }

    public struct Row: Decodable, Sendable {
        public let device: String
        public let deviceId: String
        public let cells: [Cell]
        public let usage: Usage
        public let share: Share
    }

    public struct Cell: Decodable, Sendable {
        public let usage: Usage
        public let windowTokens: Int
        public let share: Share
    }
}
